mtbl/threadpool.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2024 DomainTools LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <pthread.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <mtbl.h>
+
+#include "threadpool.h"
+
+/* Public interface */
+
+struct mtbl_threadpool *
+mtbl_threadpool_init(size_t thread_count) {
+
+	struct mtbl_threadpool *pool = calloc(1, sizeof(*pool));
+
+	if (thread_count > 0)
+		pool->pool = threadpool_init(thread_count);
+
+	return pool;
+}
+
+void
+mtbl_threadpool_destroy(struct mtbl_threadpool **poolp) {
+
+	struct mtbl_threadpool *pool = *poolp;
+	if (pool == NULL)
+		return;
+	threadpool_destroy(&pool->pool);
+	free(pool);
+	*poolp = NULL;
+}
+
+
+/* Threadpool */
+
+struct thread {
+	pthread_mutex_t m;
+	pthread_cond_t c;
+	pthread_t t;
+
+	void *arg, *res;
+
+	thread_cb cb;
+
+	struct thread *next;
+	struct resultq *rq;
+	struct threadpool *pool;
+
+	bool running;
+};
+
+struct resultq {
+	pthread_mutex_t m;
+	pthread_cond_t c;
+
+	bool finished;
+	size_t nthreads;
+
+	struct thread *head, **ptail;
+};
+
+struct threadpool {
+	pthread_mutex_t m;
+	pthread_cond_t c;
+	struct thread *head;
+
+	size_t count;
+	size_t max;
+};
+
+struct result_handler {
+	pthread_t thread;
+	struct resultq *rq;
+	result_cb cb;
+	void *cbdata;
+};
+
+static void *
```

Comment:
what's this do?

Reply:
`thread_worker()` does 2 main things: 
1. Calls the worker thread's callback function when it receives work.
2. Handles the shutdown signal for worker threads.

Reply:
please add a comment on that